File: 42-trapping-rain-water/trapping-rain-water.cpp
// optimal solution - approach
class Solution { 
public:
    int trap(vector<int>& height) {
        int water = 0;
        int leftMax = 0, rightMax = 0;
        int left = 0, right = height.size() - 1;
        while (left <= right) { // when left > right break loop as water is trapped successfully whereever possible
            if (height[left] <= height[right]) { // checks if there exists a right bar taller than this left bar
                leftMax = max(leftMax, height[left]); 
                water += leftMax - height[left]; 
                left++;
            }
            else { // if height[right] < height[left] (checks if there exists a left bar taller than this right bar)
                rightMax = max(rightMax, height[right]);
                water += rightMax - height[right];
                right--;
            }
        }
        return water;

    }
};
// T.C = O(n)
// T.C = O(1)

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
       int n = height.length;
       int leftMax = height[0];
       int[] rightMax = new int[n];
       int water = 0;
       rightMax[n - 1] = height[n - 1];
       for (int i = n - 2; i >= 0; i--) {
        rightMax[i] = Math.max(rightMax[i + 1], height[i]);
       }
       for (int i = 0; i < n; i++) {
            leftMax = Math.max(leftMax, height[i]);
            if (height[i] < leftMax && height[i] < rightMax[i]) {
                water += (Math.min(leftMax, rightMax[i]) - height[i]);
            }
       }
       return water;
    }
}
